core/src/main/java/org/infinispan/statetransfer/StateTransferInterceptor.java

```diff
@@ -186,20 +187,37 @@ public void init(StateTransferManager stateTransferManager) {
          while (ce instanceof RemoteException) {
             ce = ce.getCause();
          }
-         if (!(ce instanceof OutdatedTopologyException) && !(ce instanceof SuspectException))
-            throw throwable;
-
-         // We increment the topology id so that updateTopologyIdAndWaitForTransactionData waits for the next topology.
-         // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
-         if (trace)
-            log.tracef("Retrying command because of topology change, current topology is %d: %s",
+         if (ce instanceof OutdatedTopologyException) {
+            if (trace)
+               log.tracef("Retrying command because of topology change, current topology is %d: %s",
                   currentTopologyId(), rCommand);
-         FlagAffectedCommand flagAffectedCommand = (FlagAffectedCommand) rCommand;
-         int newTopologyId = Math.max(currentTopologyId(), flagAffectedCommand.getTopologyId() + 1);
-         flagAffectedCommand.setTopologyId(newTopologyId);
-         waitForTopology(newTopologyId);
-
-         return visitReadCommand(rCtx, flagAffectedCommand, consistentHashUpdater);
+            // We increment the topology id so that updateTopologyIdAndWaitForTransactionData waits for the next topology.
+            // Without this, we could retry the command too fast and we could get the OutdatedTopologyException again.
+            FlagAffectedCommand flagAffectedCommand = (FlagAffectedCommand) rCommand;
+            int newTopologyId = Math.max(currentTopologyId(), flagAffectedCommand.getTopologyId() + 1);
+            flagAffectedCommand.setTopologyId(newTopologyId);
+            CompletableFuture<Void> future = stateTransferLock.topologyFuture(newTopologyId);
+            if (future != null) {
+               return future.thenCompose(ignored -> {
+                  try {
+                     return visitReadCommand(rCtx, flagAffectedCommand, consistentHashUpdater);
+                  } catch (Throwable t) {
+                     throw new CacheException(t);
+                  }
+               });
+            }
+         } else if (ce instanceof SuspectException) {
+            if (trace)
+               log.tracef("Retrying command because of suspected node, current topology is %d: %s",
+                  currentTopologyId(), rCommand);
+            // Upon SuspectException we don't wait for newer topology, as it is possible that current topology
+            // is actual but the view still contains a node that's about to leave; a broadcast to all nodes
+            // then can end with suspect exception, but we won't get any new topology. An example of this situation
+            // is when a node sends leave - topology can be installed before the new view.
```

Comment:
Non-broadcast RPCs can also be sent as broadcasts if the cluster has more than 2 nodes and all the nodes are in the recipients list. Say if you have a distributed cache running on 2 nodes and you start a 3rd, as long as the rebalance is in progress you will have some keys with 3 owners and a write to them will be sent as a broadcast.

I don't think it's ok to retry immediately on a SuspectException, we should always wait for something. Here we can check whether `ce.getSuspect()` is still a member of the topology, and retry immediately only if it's not a member. That way, when a node crashes, we won't use 100% of a CPU (or maybe the only CPU!) retrying the same command until we receive the new topology.

Maybe we should also extend this to throw an exception if the current topology id is the same as the command's topology id AND the suspected node is not a member in the topology -- that probably means the command is replicated using `SYNCHRONOUS` when it should be using `SYNCHRONOUS_IGNORE_LEAVERS`. Or maybe I should create a separate JIRA :)